[Validator] Add capability to use callable object (```__invoke()``` method) as callback validator

ad3n opened the pull request:
Add capability to call __invoke(callable object) as validator callback

| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no   
| Deprecations? | no 
| License       | MIT

Add capability to use callable object as callback validator

Before:

```@Assert\Callback({"App\Validator\YourCustomValidator", "__invoke"})```

After:

```@Assert\Callback("App\Validator\YourCustomValidator")```

ad3n committed:
Add capability to call __invoke method

Add capability to call __invoke(callable object) as validator callback

yceruto commented on src/Symfony/Component/Validator/Constraints/CallbackValidator.php:
```diff
@@ -44,8 +44,11 @@ public function validate($object, Constraint $constraint)
 
             \call_user_func($method, $object, $this->context, $constraint->payload);
+        } elseif (\method_exists($method, '__invoke')) {
+            $validator = new $method();
```
what if `$method` is an invokable object? We need to check it before building a new instance, no?

ad3n replied:
when the ```__invoke``` method is exist, we imagine the object is callable, and we need to create an object before because the ```__invoke``` only available in object

ad3n replied:
@yceruto: in validator, you only pass the string of class name or a method name of object (entity), you can't passing object inside of validator (annotation) except it self (entity)

ogizanagi replied:
Actually you can in some ways, using PHP. But that's probably discouraged as the constraint needs to be serializable to be cached, so could be an issue/not work as expected. A custom constraint & validator should be preferred.

ad3n replied:
@ogizanagi : can you tell how to fix?

ostrolucky commented on src/Symfony/Component/Validator/Constraints/CallbackValidator.php:
```diff
@@ -44,8 +44,11 @@ public function validate($object, Constraint $constraint)
             }
 
             \call_user_func($method, $object, $this->context, $constraint->payload);
+        } elseif (\method_exists($method, '__invoke')) {
```
Instead of adding this branch, previous one should be modified
```diff
- if ($method instanceof \Closure) {
+ if (is_callable($method)) {
```

ad3n replied:
when we use ```\Closure``` and checking use ```is_callable()```, the result is always false in this case. Because ```$method``` is string contain the class name.

ad3n replied:
You can check, when we use ```is_callable()``` we get an exception ```Method "%s" targeted by Callback constraint does not exist in class %s```

ostrolucky replied:
Well it shouldn't be a string, my guess is you use it incorrectly. $method should be invokable, not a method of new instance of $method.

ad3n replied:
@ostrolucky : i added a test to check if ```$method``` is object or not, can you check the new one?

ad3n committed:
check ```$method``` is object

check ```$method``` is object, if it false, we create an object then invoked

ad3n committed:
Fix coding standard

ad3n committed:
Trap the method if method is invokable object

This commit make sure ```$method``` has ```__invoke``` method and ```$method``` is not callable object.

xabbuh commented:
I don't see how this would actually work. When using `@Assert\Callback("App\Validator\YourCustomValidator")`, the validator will not receive an object of that class, but only the class name as a string.

ad3n commented:
@xabbuh : Yes, you are right, that why we need to create an object before calling the validator

@https://github.com/ad3n/symfony/blob/patch-1/src/Symfony/Component/Validator/Constraints/CallbackValidator.php#L48

xabbuh commented:
I am not convinced adding such a feature would be a good idea. Whether or not we will be able to create the object needed to be able to call `__invoke()` would completely depend on the constructor signature. From my point of view the risk for users breaking their application by changing the constructor signature is too high. Especially given the fact that you would only notice these failures at runtime.

That being said I am 👎 on adding this feature. IMO for this you should rather stick with a custom constraint validator.

fabpot commented:
I agree with @xabbuh, thanks for suggesting.
